### ISSUE TITLE
`$fetch` cannot fetch files in `public/` directory isomorphically

### ISSUE THREAD

**Issue by DaniFoldi** [nuxt/nuxt#13857]
### Environment

Nuxt CLI v3.0.0-rc.1
RootDir: /home/projects/github-sn7zlr
Nuxt project info:

------------------------------
- Operating System: `Linux`
- Node Version:     `v16.14.2`
- Nuxt Version:     `3.0.0-rc.1`
- Package Manager:  `npm@7.17.0`
- Builder:          `vite`
- User Config:      `-`
- Runtime Modules:  `-`
- Build Modules:    `-`
------------------------------

### Reproduction

https://stackblitz.com/edit/github-sn7zlr?file=pages%2Ftest.vue

### Describe the bug

- Visit `/` and click on the link to go to `/test` -> json data shows normally
- Visit `/test` directly -> json data is null

**Comment by semiaddict** [comment]
I am unable to load files from the public directory at all from the client side.

I have tried loading a simple json from the public directory using $fetch and useFetch from within a component's setup function, and a server API handler, they both give me 500 errors.

I see the label `workaroud available` has been added to this issue, but can't see any listed here. What is the workaround?

**Comment by semiaddict** [comment]
The $fetch in the component setup seems to work if I use it with `then`, but not with `await` even if the setup function is declared as async. I guess this is then another issue.

However, the $fetch inside an API handler only works if the URL is absolute; so `await $fetch("/test.json")` doesn't work, while `await $fetch("http://localhost:3000/test.json")` does.

**Comment by vdilly** [comment]
@semiaddict indeed your workaround works using $fetch with an absolute URL to the json file !

However that doesn't work locally for me. Using npm run generate in local developement environment, the routes containing the fetch to a .json file are not rendered.

**Comment by semiaddict** [comment]
@vdilly, what are you trying to do exactly ?
I personally needed to create a pre-rendered route that generated a search index based on content.
I ended up with something like this in a module which works great for my use case:

**modules/search/index.js:**
```ts
export default defineNuxtModule<ModuleOptions>({
   //...
  async setup (moduleOptions, nuxt) {
    const { resolve } = createResolver(import.meta.url);

    // Add the search index route.
    addServerHandler({
      method: 'get',
      route: "/search-index.json",
      handler: resolve('./runtime/server/api/search-index'),
    });

    // Pre-render the search index route for SSG.
    addPrerenderRoutes("/search-index.json");
  },
});
```

**modules/runtime/server/api/search-index.ts:**
```ts
import { defineEventHandler } from 'h3';

// This route is used to generate the search index.
export default defineEventHandler(async (event) => {
  const index = {};
  // ... generate the JSON data
  return JSON.stringify(index);
})
```

Then in your component's setup you should be able to call:
```ts
const { data } = await useFetch("/search-index.json");
```

**Comment by vdilly** [comment]
@semiaddict thanks for taking the time to share this ! 
I didn't take a good look at nuxt modules yet, but I definetly gonna try the way you handled that.

What I did was not very "nuxty", 
I simply created a node script that fetch the datas I needed, write them into multiple chunk JSON files in /public folder. 
I put this script in the "pregenerate" scripts of my package.json so it's executed before nuxt generate.
Then in my components setup, I $fetch the json I need with an absolute URL (wich works great on Netlify for example but not in my local environment for some reasons)

The goal was to reduce the build time, caching the datas that a lot of my pages shares and prevents fetching my api for the same datas over hundreds of routes.

**Comment by semiaddict** [comment]
> I simply created a node script that fetch the datas I needed, write them into multiple chunk JSON files in /public folder.
I put this script in the "pregenerate" scripts of my package.json so it's executed before nuxt generate.
Then in my components setup, I $fetch the /public/[name].json I need with an absolute URL (wich works great on Netlify for example but not in my local environment for some reasons)

Then I guess you could have the _pregenerate_ script write the JSONs to a module's directory such as "modules/mymodule/runtime/assets/json/" and then use the same technique I mentioned above by defining pre-rendered server routes to serve those JSON files.

I wonder, however, if you really need to get into all this trouble if you are building with SSG, as I believe if you correctly use [useAsyncData](https://nuxt.com/docs/api/composables/use-async-data) with a unique identifier per api along with [$fetch](https://nuxt.com/docs/api/utils/dollarfetch#fetch), the identical api endpoint will be fetched only once for all components; but I haven't actually verified this.

**Comment by vdilly** [comment]
For what I experienced, this is true with SSR/CSR but not across different routes with SSG. I think Nuxt might treat each route as single instance and so the cache related to the "key" of useFetch/UseAsyncData is not shared between routes (this still might be true between multiple components of the same route though).
I might be missing something though

I'll try what you suggest for the json file and give a feedback, thanks !

**Comment by chris-paganon** [comment]
From what I see, the root cause is that the node server is running from `.output/server/` but the public folder is at `.output/public`
So when we want to access a file from the server that would be in the public folder and we run `await $fetch("/test.json")`, we actually access this folder `.output/server/test.json` instead of `.output/public/test.json`.

For now the full URL workaround will do. I use a runtimeConfig variable to define the URL based on the environment and it works alright, although not ideal.

**Comment by ogios** [comment]
any progress?

**Comment by Skyost** [comment]
Anything new here ?

**Comment by stephanelgrg** [comment]
Hello, it seems to be working now with nuxt@3.21.9
https://stackblitz.com/edit/github-sn7zlr-gxrstjmr?file=pages%2Ftest.vue,nuxt.config.ts

<img width="1612" height="315" alt="Image" src="https://github.com/user-attachments/assets/53205b6f-d64f-4fca-a4da-5e7a3a3571d2" />

Have a good day 🌞 

@danielroe 🫶

**Comment by danielroe** [comment]
it works in dev mode but not when deployed or prerendering, I believe.

**Comment by lutejka** [comment]
When I was checking this yesterday it was also not working for me in dev mode 

<img width="1496" height="709" alt="Image" src="https://github.com/user-attachments/assets/b7e88a16-cef1-4bbc-8551-f4fac7ba289a" />

Also with https://stackblitz.com/edit/github-sn7zlr-gxrstjmr?file=pages%2Ftest.vue,nuxt.config.ts from stephan, the json can not be fetch when i open /test directly
